lib/ui/painting/image_decoder.cc

```diff
@@ -216,38 +216,55 @@ static SkiaGPUObject<SkImage> UploadRasterImage(
   return result;
 }
 
-void ImageDecoder::Decode(fml::RefPtr<ImageDescriptor> descriptor,
+void ImageDecoder::Decode(fml::RefPtr<ImageDescriptor> descriptor_ref_ptr,
                           uint32_t target_width,
                           uint32_t target_height,
                           const ImageResult& callback) {
   TRACE_EVENT0("flutter", __FUNCTION__);
   fml::tracing::TraceFlow flow(__FUNCTION__);
 
+  // ImageDescriptors have Dart peers that must be collected on the UI thread.
+  // However, closures in MakeCopyable below capture the descriptor. The
+  // captures of copyable closures may be collected on any of the thread
+  // participating in task execution.
+  //
+  // To avoid this issue, we resort to manually reference counting the
+  // descriptor. Since all task flows invoke the `result` callback, the raw
+  // descriptor is retained in the beginning and released in the `result`
+  // callback.
+  //
+  // `ImageDecoder::Decode` itself is invoked on the UI thread, so the
+  // collection of the smart pointer from which we obtained the raw descriptor
+  // is fine in this scope.
+  auto raw_descriptor = descriptor_ref_ptr.get();
```

Comment:
It might be useful to add a method to `fml::RefPtr` that detaches the raw pointer from the `RefPtr` and gives the caller ownership of the ref count that was held by the `RefPtr`.

(similar to `std::unique_ptr::release` or Skia's `sk_sp::release`)